### PR TITLE
Invite notification fixes #959

### LINKED ISSUE #595: Signups "comment" field is string instead of text
People leaving long comments gets it all truncated because of the wrong type for the column. Migrate.

### LINKED ISSUE #959: NeuroHub Alpha user feedback - technical issue - user invitation and notification
Received via direct email message sent to Diana:

a) Invite users: easy, I invited Diana, invitation is pending. I would like to receive a notification when the person ‘accept’ my invitation.

### PR DESCRIPTION
**Pull request by MontrealSergiy**
This adds notification for invitation acceptance using elements of cbrain messaging system.

### CODE REVIEW

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb [hunk 1]
```diff
@@ -35,3 +35,16 @@
       <%= (h(msg.to_s.strip).gsub(/\n/, '<BR/>')).html_safe %>
     </div>
   <%- end -%>
+
+  <div class="card">
```
Comment (prioux) [inline BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:39]: Shouldn't the div be inside the `if` block ? Otherwise we get a div with nothing inside.

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.  
+#
+-%>
+<% messtype = mess.message_type # disguise confirmation for a flash notice until better styling added
+   prefix = "flash"
+   messclass = "#{prefix}_#{messtype}"
+%>
+
+<% unless mess.description.blank? %>
```
Comment (prioux) [inline BrainPortal/app/views/nh_invitations/_message_details.html.erb:29]: Change the `unless something.blank?` to `if something.present?` it looks much nicer and avoids double negatives.

File: BrainPortal/app/controllers/invitations_controller.rb [hunk 3]
```diff
@@ -105,6 +105,14 @@ def update #:nodoc:
     @invitation.save
 
     flash[:notice] = "You have been added to project #{@group.name}."
+
+    Message.send_message(@invitation.sender,
+                         :message_type   => 'notice',
+                         :header         => "Invitation Accepted",
+                         :description    => "A user joined project #{@nh_invitation.group.name}",
```
Comment (natacha-beck) [inline BrainPortal/app/controllers/invitations_controller.rb:112]: I confirm it make it crash.

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb (hunk 1 above)
Comment (natacha-beck) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:39]: it should be inside the `unless` block.

Comment (natacha-beck) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:39]: + same comment as @prioux made later:
> Change the unless something.blank? to if something.present? it looks much nicer and avoids double negatives.

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb [hunk 4]
```diff
@@ -35,3 +35,16 @@
...
+  <%- unless @display_messages.blank? %>
+
+    <%- for mess in @display_messages -%>
+      <p>
+      <div class="nh-invitations card-item">
+      <%= render(:partial  => "nh_invitations/message_details", :locals  => { :mess  => mess } ) rescue "Oh oh. your message was not rendered. Talk to devs!" %>
```
Comment (natacha-beck) [inline BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: You do not need a rescue on a render. Rails will handle that.

In meantime we have the same code in `./messages/_message_display.html.erb` ==> ligne 33.

Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: well, much code is similiar to similiar cbrain features. Do you say I should 'unrescue' old code too?

File: BrainPortal/app/controllers/invitations_controller.rb [hunk 5]
```diff
@@ -105,6 +105,14 @@ def update #:nodoc:
     @invitation.save
 
     flash[:notice] = "You have been added to project #{@group.name}."
+
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/invitations_controller.rb:108]: I guess @group will be best or at the least the shortest

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb (hunk 4 above)
Comment (natacha-beck) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: Can you give me example in CBRAIN ? I did not see them.

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 6]
```diff
@@ -83,6 +84,36 @@ def check_if_locked
...
+  ########################################################################
+  # Messaging System Filters (presently only invite acceptance)
+  ########################################################################
+
+  # Find new messages to be displayed at the top of the page.
+  def prepare_messages #:nodoc:
+    return unless current_user
+    return if     current_user.all_licenses_signed.blank?
+    return if     request.format.blank? || request.xhr?
+    return unless request.format.to_sym == :html || params[:controller] == 'messages'
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:96]: well... I think putting && it will break the functionality (mind `unless`). As a compromise I can remove or comment out  the second part of the condition since, at the moment, there is no messages controller in NeuroHub part at the moment.

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 2 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:29]: ok Not sure is being blank something negative, `if` is shorter and as I know many devs consider `unless` bad style unless it removes a double negation.

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb (hunk 4 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: ./messages/_message_display.html.erb ==> ligne 33.

Comment (natacha-beck) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: Yep, I already mention it. Seems to be the 2 only place with this strange rescue. Please remove at both place.

Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb:45]: besides the templates, much of message processing in neurohub_controller.rb is copy of cbrain_controller.

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 7]
```diff
@@ -83,6 +84,36 @@ def check_if_locked
...
+
+    @display_messages = []
+
+    unread_messages = current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
+    @unread_message_count = unread_messages.count
```
[the remark this answers is not in the document]
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:101]: no reason. The possible future use is to put number of the unread messages over 'message' item of the navigation bar/top menu like Candice done. I guess can be deleted both from CBRAIN and NeuroHub controllers.

[the remark this answers is not in the document]
Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:101]: Don't do that. Just keep the code clean. If someone in the future needs an instance variable, they'll add it then.

File: BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb (hunk 4 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: It does not handle well errors on my pc image

Comment (prioux) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: Try in the Rails `production` environment

Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: in production mode during malfunction the message seems silently hidden, which mean the user will do not know that his message got lost. Do you still want remove that rescues ? @natacha-beck

Comment (prioux) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: Just use the same `rescue` that was there on the CBRAIN side, then.

Comment (prioux) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: I don't want to waste too much time about this anymore.

Comment (MontrealSergiy) [reply BrainPortal/app/views/layouts/_neurohub_flash_message.html.erb]: As a compromise I removed the rescue form neurohub as users might understand that it is experimental project and confirmations could be lost

**Comment by MontrealSergiy** [comment]
@natacha-beck actually you are right the rails handles templates error and produces an error message saying admins are notified of the issue. Regretfully for some reason when that happens neurohub redirects user to cbrains group page which seems strange.

Screenshot 17 08 2020 22_47_19

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 8]
```diff
@@ -31,6 +31,7 @@ class NeurohubApplicationController < ApplicationController
 
   before_action :switch_to_neurohub_layout
   before_action :prepare_invites
+  before_action :prepare_messages
```
Comment (prioux) [inline BrainPortal/app/controllers/neurohub_application_controller.rb:34]: Now you're asking the rails framework to call `prepare_messages` twice for every request, because the callback is already registered in ApplicationController

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 9]
```diff
@@ -83,6 +84,35 @@ def check_if_locked
 
   end
 
+  ########################################################################
+  # Messaging System Filters (presently only invite acceptance)
+  ########################################################################
+
+  # Find new messages to be displayed at the top of the page.
+  def prepare_messages #:nodoc:
```
Comment (prioux) [inline BrainPortal/app/controllers/neurohub_application_controller.rb:92]: This is a complete duplication of the method of the same name in Application controller, except for a single line. That is not DRY at all.

Please:

1. remove this method here since it's already defined in ApplicationController
2. change the method in ApplicationController such that the one line that differs invokes a method that can be overrided, which you will add to both files:
```ruby
   unread_messages = unread_messages_to_display
```

so that  in ApplicationController:

```ruby
# put this method just after prepare_messages
def unread_messages_to_display 
   current_user.messages.where( :read => false ).order( "last_sent DESC" )
end
```

and in NeurohubApplicationController we ONLY have this now:

```ruby
def unread_messages_to_display
  current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
end
```

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb [hunk 10]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (prioux) [inline BrainPortal/app/views/nh_invitations/_message_details.html.erb:1]: This file seems like aalso a almost complete duplication of `messages/_message_details.html.erb`, can we instead use THAT and not duplicate so much code?

File: BrainPortal/app/controllers/neurohub_application_controller.rb (hunk 8 above)
Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:34]: I checked and it's not called twice, Rails is intelligent enough not to invoke a method twice if you register it several time. But the point is still valid, it's already registered in ApplicationController.

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 11]
```diff
@@ -83,6 +84,35 @@ def check_if_locked
...
+    return unless current_user
+    return if     current_user.all_licenses_signed.blank?
+    return if     request.format.blank? || request.xhr?
+    return unless request.format.to_sym == :html
+
+    @display_messages = []
+
+    unread_messages = current_user.messages.where( :read => false, :header => 'Invitation Accepted' ).order( "last_sent DESC" )
```
Comment (prioux) [inline BrainPortal/app/controllers/neurohub_application_controller.rb:100]: This query is brittle by the way. If someone in the future change the header of the message that is being sent (see your code in InvitationControlelr and NhInvitationController), then this piece of code will stop displaying the message.

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.  
+#
+-%>
+<% messtype = mess.message_type
+   prefix = "flash"  # disguise an invite confirmation message for a flash notice until proper styling defined
+   messclass = "#{prefix}_#{messtype}"
+%>
+
+<% if mess.description.present? %>
+  <p class=<%= messclass %>>
```
Comment (prioux) [inline BrainPortal/app/views/nh_invitations/_message_details.html.erb:30]: By the way, the values of the HTML attributes should be quoted. They were quoted in the original `messages/_message_details.html.erb`, why did you remove the quotes in your copy?

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 10 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:1]: Surely, it is easy to adjust that file. That what I done in an early version of the feature. I opted to allow Neurohub have it own template that can evolve independtly, but totally can drop that

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 12 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:30]: I tried a bit improve over existing code.

Comment (MontrealSergiy) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:30]: Note If I add quotes to this version I will get double quotes

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 10 above)
Comment (prioux) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:1]: OK, I agree with this sentiment. Let's have a separate file for neurohub, like you did.

Comment (prioux) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:1]: You convinced me.

File: BrainPortal/app/controllers/neurohub_application_controller.rb (hunk 11 above)
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:100]: Should I then add `message_type=confirmation` flag along with relevant styles? 

Showing all the cbrain messages seems a bit risky. Or as a half measure may be add a unittest?

Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:100]: Just use the same query but encspsulate it in a method, like I explained above, so that you only have one 'prepare_message' callback in ApplicationController.

File: BrainPortal/app/controllers/nh_invitations_controller.rb [hunk 13]
```diff
@@ -132,6 +132,13 @@ def update #:nodoc:
     @nh_invitation.save
 
     flash[:notice] = "You have been added to project #{@nh_project.name}."
+
+    Message.send_message(@nh_invitation.sender,
```
Comment (prioux) [inline BrainPortal/app/controllers/nh_invitations_controller.rb:136]: By the way, here, why not use @nh_project.name instead of @nh_imvitation.group.name ?

Comment (MontrealSergiy) [reply BrainPortal/app/controllers/nh_invitations_controller.rb:136]: ok that's easy. I experimented a bit to make cut and paste from cbrain which we often use for neurohub smoother ( you do not have to remember that groups became nh_projects, just add/remove nh). But can change.

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 14]
```diff
@@ -122,5 +97,13 @@ def prepare_invites
     @nh_new_invites_count = nh_new_invites.count
   end
 
+  # this patches prepare_message method
+  # to display 'Invitation Accepted' entitled messages. They will be show separately from Invitations
+  # and only kind of message to be shown for now.
+
```
Comment (prioux) [inline BrainPortal/app/controllers/neurohub_application_controller.rb:103]: You must not leave a blank line between the method description and the method, otherwise the documetnation code produces a method with no description

Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:103]: I'd also suggest adding the #:nodoc: keyword because that method is not really a method that CBRAIN developer will want to invoke at other places.

Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:103]: Same with the copy in ApplicationController

File: BrainPortal/app/controllers/neurohub_application_controller.rb (hunk 8 above)
Comment (prioux) [reply BrainPortal/app/controllers/neurohub_application_controller.rb:34]: Please remove it

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 12 above)
Comment (prioux) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb:30]: This is not fixed yet

File: BrainPortal/app/controllers/neurohub_application_controller.rb (hunk 9 above)
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/neurohub_application_controller.rb]: added

File: BrainPortal/app/views/nh_invitations/_message_details.html.erb (hunk 12 above)
Comment (MontrealSergiy) [reply BrainPortal/app/views/nh_invitations/_message_details.html.erb]: added

File: BrainPortal/app/controllers/neurohub_application_controller.rb (hunk 14 above)
Comment (MontrealSergiy) [reply BrainPortal/app/controllers/neurohub_application_controller.rb]: nodoc added

**Comment by MontrealSergiy** [comment]
I think I addressed the remaining comments. I kept a query method description separated because, as you noticed, it is fragile but put the nodoc thing.

File: BrainPortal/app/controllers/neurohub_application_controller.rb [hunk 15]
```diff
@@ -91,6 +95,12 @@ def prepare_invites
     @nh_new_invites_count = nh_new_invites.count
   end
 
+  # below a method is overwriten to render only invitation confirmation, filtered by header (for now)
+
```
Comment (prioux) [inline BrainPortal/app/controllers/neurohub_application_controller.rb:99]: Don't leave a blank line!
